src/datasets/arrow_dataset.py

```diff
@@ -3358,6 +3358,468 @@ def init_buffer_and_writer():
+                 If `batched=True`, the initializer must be set.
+             batched (`bool`, defaults to `False`):
+                 Provide batch of examples to `function`.
+             batch_size (`int`, *optional*, defaults to `1000`):
+                 Number of examples per batch provided to `function` if `batched=True`.
+                 If `batch_size <= 0` or `batch_size == None`, provide the full dataset as a single batch to `function`.
+             drop_last_batch (`bool`, defaults to `False`):
+                 Whether a last batch smaller than the batch_size should be
+                 dropped instead of being processed by the function.
+             keep_in_memory (`bool`, defaults to `False`):
+                 Keep the dataset in memory instead of writing it to a cache file.
+             load_from_cache_file (`bool`, defaults to `True` if caching is enabled):
+                 If a cache file storing the current computation from `function`
+                 can be identified, use it instead of recomputing.
+             cache_file_name (`str`, *optional*, defaults to `None`):
+                 Provide the name of a path for the cache file. It is used to store the
+                 results of the computation instead of the automatically generated cache file name.
+             fn_kwargs (`Dict`, *optional*, defaults to `None`):
+                 Keyword arguments to be passed to `function`.
+             num_proc (`int`, *optional*, defaults to `None`):
+                 Max number of processes when generating cache. Already cached shards are loaded sequentially.
+             new_fingerprint (`str`, *optional*, defaults to `None`):
+                 The new fingerprint of the dataset after transform.
+                 If `None`, the new fingerprint is computed using a hash of the previous fingerprint, and the transform arguments.
+             desc (`str`, *optional*, defaults to `None`):
+                 Meaningful description to be displayed alongside with the progress bar while mapping examples.
+
+         Example:
+
+         ```py
+         >>> from datasets import load_dataset
+         >>> from collections import Counter
+
+         # Count the number of words in the validation set of the Rotten Tomatoes dataset
+         >>> rotten_ds = load_dataset("rotten_tomatoes", split="validation")
+         >>> count = lambda counter, text: counter + Counter(text.lower().split())
+         >>> sum_counts = lambda counter1, counter2: counter1 + counter2
+         >>> result = result = rotten_ds.select(range(1)).reduce(count, initializer=Counter(), input_columns="text")
+         >>> result
+        {'text': Counter({'and': 2, 'compassionately': 1, 'explores': 1, 'the': 1, 'seemingly': 1, 'irreconcilable': 1, 'situation': 1, 'between': 1, 'conservative': 1, 'christian': 1, 'parents': 1, 'their': 1, 'estranged': 1, 'gay': 1, 'lesbian': 1, 'children': 1, '.': 1})}
```

Comment:
Since the initializer is a `Counter()`, I think we can expect the output to be a `Counter()` as well no ?

Reply:
I see your point! It would also align more with the behaviour of `functools.reduce`. On the other hand, its a bit annoying that the output of different parameter variations is not the same type, as an example, `r1` and `r2` below would be `int` and `dict` respectively.
```python
int_ds = Dataset.from_dict({"x": [1, 2, 3], "y": [1, 2, 3]})
r1 = reduce(function, initializer=0)
r2 = reduce(function)
```
I don't really have a strong opinion one way or the other, either would be confusion/annoying in some way. Which do you prefer?